osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs

```diff
@@ -51,33 +75,52 @@ public override double Calculate(Dictionary<string, double> categoryRatings = nu
             countMeh = Score.Statistics.GetOrDefault(HitResult.Meh);
             countMiss = Score.Statistics.GetOrDefault(HitResult.Miss);
 
+            greatWindow = 79.5 - 6 * Attributes.OverallDifficulty;
+
             // Don't count scores made with supposedly unranked mods
             if (mods.Any(m => !m.Ranked))
                 return 0;
 
             // Custom multipliers for NoFail and SpunOut.
-            double multiplier = 1.12; // This is being adjusted to keep the final pp value scaled around what it used to be when changing things
+            double multiplier = 2.14; // This is being adjusted to keep the final pp value scaled around what it used to be when changing things
 
             if (mods.Any(m => m is OsuModNoFail))
                 multiplier *= 0.90;
 
             if (mods.Any(m => m is OsuModSpunOut))
                 multiplier *= 0.95;
 
+            // guess the number of misses + slider breaks from combo
+            double comboBasedMissCount;
+
+            if (countSliders == 0)
+            {
+                if (scoreMaxCombo < beatmapMaxCombo)
+                    comboBasedMissCount = (double)beatmapMaxCombo / scoreMaxCombo;
+                else
+                    comboBasedMissCount = 0;
+            }
+            else
+            {
+                double fullComboThreshold = beatmapMaxCombo - 0.1 * countSliders;
+                if (scoreMaxCombo < fullComboThreshold)
+                    comboBasedMissCount = fullComboThreshold / scoreMaxCombo;
+                else
+                    comboBasedMissCount = Math.Pow((beatmapMaxCombo - scoreMaxCombo) / (0.1 * countSliders), 3);
```

Comment:
Is this approximation derived in some way, or was it just an experimentation-based "good enough" one?

Reply:
if there were no slider ends, and if you had a combo of x out of max, then you must have at least ceil(max/x) slider breaks. It uses a continuous value instead of ceil because it's weird for PP to step discretely at certain combos, and it eases off towards the end because slider ends don't count.